cranelift/codegen/src/isa/x64/inst/encoding/rex.rs

```diff
@@ -0,0 +1,453 @@
+//! Encodes instructions in the standard x86 encoding mode. This is called IA-32E mode in the Intel
+//! manuals but corresponds to the addition of the REX-prefix format (hence the name of this module)
+//! that allowed encoding instructions in both compatibility mode (32-bit instructions running on a
+//! 64-bit OS) and in 64-bit mode (using the full 64-bit address space).
+//!
+//! For all of the routines that take both a memory-or-reg operand (sometimes called "E" in the
+//! Intel documentation, see the Intel Developer's manual, vol. 2, section A.2) and a reg-only
+//! operand ("G" in Intelese), the order is always G first, then E. The term "enc" in the following
+//! means "hardware register encoding number".
+
+use crate::{
+    ir::TrapCode,
+    isa::x64::inst::{
+        args::{Amode, OperandSize},
+        regs, EmitInfo, EmitState, Inst, LabelUse,
+    },
+    machinst::{MachBuffer, MachInstEmitInfo},
+};
+use regalloc::{Reg, RegClass};
```

Comment:
In a future change, I would like to:
 - factor out the external dependencies to this encoding code (I mean, why not? It would make it a bit easier to test and perhaps this code could be useful elsewhere?)
 - implement a few more abstractions like `RexFlags` to make this easier to work with (again, why not? Rust == "zero-cost abstractions," right? 😀)

Any thoughts for or against this?

Reply:
+1 to both of these!

In particular to the first point: there is really an x86-64 assembler library begging to escape from this module -- it is "incomplete" in the sense that it only implements the instructions we actually need, but maybe it could be useful for other folks in the Rust world as well, eventually.